(PDB-2387) Add config flags for SSL auth

ajroetker opened the pull request:
This commit allows import and export arguments (specifically infile and
outfile respectively) to be supplied positionally to at the command
line. Whereas before a user would have needed to do

```
    puppet-db import --infile foo
```

They may now do

```
    puppet-db import foo
```

This PR also adds `--urls`, `--cacert/--cert/--key` flags to the global options for the CLI and now makes sure we parse `server_urls` when it's a string.

mullr commented on exe/puppet-db.cc:
```diff
@@ -19,13 +19,14 @@ namespace logging = leatherman::logging;
-                 << global_desc << "\n"
-                 << export_subcommand_desc << "\n"
-                 << import_subcommand_desc << endl;
+    os << "usage: puppet-db [global] export [options]\n"
```
\n => endl

kbarber commented on exe/puppet-db.cc:
```diff
@@ -40,6 +41,16 @@ main(int argc, char **argv) {
         po::options_description global_options("global options");
         global_options.add_options()
                 ("help,h", "produce help message")
+                ("config", po::value<string>()->default_value(""),
```
One thing to note here, long GNU options are not loved by everyone in the unix world, in fact there is a set of admins prefer the short options for succinctness. This is why most programs have support for long and short options, to appease everyone. Having to type puppet foo --config 'bar.cfg' versus puppet foo -c 'bar.cfg' for example.

kbarber commented on CMakeLists.txt:
```diff
@@ -78,9 +78,6 @@ add_cppcheck_dirs("${PROJECT_SOURCE_DIR}/lib" "${PROJECT_SOURCE_DIR}/exe")
 
-# We duplicate the `help_test` as `smoke_test` until the default functionality
-# of the tool has been nailed down.
-add_test(NAME "db_smoke_test" COMMAND puppet-db)
```
Question, how are we testing the default behaviour now?

ajroetker replied:
We'll have to do that at a higher level, since this command (`puppet-db`) is actually incomplete and so the output is directed to nowide::cerr and thus the cli returns a non-zero exit status.

kbarber commented on lib/inc/puppetdb-cli/puppetdb-cli.hpp:
```diff
@@ -28,10 +28,30 @@ class LIBPUPPETDB_EXPORT PuppetDBConn  {
+     * @param cert client cert to use for curl.
+     * @param key client private key to use for curl.
+     */
+    PuppetDBConn(const std::string& urls,
```
Positional args don't make for a great API, we've already got a couple of variants in this patch for example. Have we considered taking a struct instead?

ajroetker replied:
Then we're just wrapping things in a class to pass them to another class, this way you know the api by looking at the function and you don't have to go track down the struct

kbarber replied:
I don't think using structs is abnormal however, quite normal in a C/C++ world I would argue ... the problem you avoid is the extrapolation of extra parameters as this gets more and more complex over time. At five parameters, this is getting quite long also, and it raised alarm bells in my head. With a struct you can add fields, deprecate fields and all kinds of magic like that without affecting the API, with positional arguments this becomes harder and you can box yourself in easier.

Also, be mindful that these parameters are wrapped and passed to other functions like you say, you can pass each one individually right, or you can just pass the struct pointer.

Most of my concern is that I can see this thing becoming the pattern for an external api library if we keep going, and exposing this to other consumers concerns me as it stands.

kbarber commented on exe/puppet-db.cc:
```diff
@@ -19,13 +19,14 @@ namespace logging = leatherman::logging;
-                 << export_subcommand_desc << "\n"
-                 << import_subcommand_desc << endl;
+    os << "usage: puppet-db [global] export [options]\n"
+       << "       puppet-db [global] import [options]\n"
```
This usage doesn't really express the new positional syntax.

kbarber commented on exe/puppet-db.cc:
```diff
@@ -40,6 +41,16 @@ main(int argc, char **argv) {
+                ("config", po::value<string>()->default_value(""),
+                 "path to use for CLI configuration")
+                ("urls", po::value<string>()->default_value(""),
+                 "urls to use for PuppetDB")
```
You might want to mention how one can provide multiple values.

kbarber commented on lib/src/puppetdb-cli.cc:
```diff
@@ -31,49 +33,111 @@ version()
+            "~/.puppetlabs/client-tools/puppetdb.conf" : config_path;
     const string expanded_config_path
-    { futil::tilde_expand("~/.puppetlabs/client-tools/puppetdb.conf") };
+    { futil::tilde_expand(defaulted_config_path) };
```
Why is this default expressed late, and not in the main options area? I can't see what the default is when I do a -h atm for example, might be nice.

kbarber commented on lib/src/puppetdb-cli.cc:
```diff
@@ -31,49 +33,111 @@ version()
+                           const string& cert,
+                           const string& key) :
+        server_urls_ { urls.empty() ?
+            vector<string>({ "http://127.0.0.1:8080" }) :
```
Why isn't this just a default in the options?

ajroetker replied:
Then we can't easily say "use `--urls`, if that's empty, use `--config`"

kbarber replied:
Huh, bummer. I guess wanting I'm really looking for is the help output showing what the default is, right now it shows no default because its here. Maybe our way of documenting defaults is the wrong pattern or something ... hrm.

ajroetker replied:
We could put it in the configuration file documentation. Or we could put it in the doc for the `-c,config` flag. Thoughts?

kbarber replied:
Either/or ... :-).

kbarber commented on lib/src/puppetdb-cli.cc:
```diff
@@ -31,49 +33,111 @@ version()
     return puppetdb_conn;
 }
 
+vector<string> split(string str, char delimiter) {
```
Any reason we can't use boost::split from boost/algorithm/string.hpp?

kbarber commented on lib/src/puppetdb-cli.cc:
```diff
@@ -31,49 +33,111 @@ version()
+
+server_urls_t
+parse_server_urls_str(const string& urls) {
+    return split(urls, ',');
```
How sensitive is this mechanism, to spaces before/after the comma? Like if I quote a string on the cli for example, and do this what happens?

ajroetker replied:
Everything behaves correctly, it's not sensitive to spaces

kbarber commented on exe/puppet-query.cc:
```diff
@@ -60,7 +70,7 @@ main(int argc, char **argv) {
 
             if (vm.count("help")) {
-                help(global_options);
+                help(global_options, nowide::cout);
```
This made me pause, but the cout behaviour is correct when people ask for help, a comment to this affect reminding everyone might stop people from pausing :-).

kbarber commented:
@ajroetker btw, you forgot to document the new configuration items in the two pod's I think (unless I'm missing something).

Worth noting I can't see what a configuration file should look like in either pod file.

kbarber commented:
@ajroetker oh yeah, how do we test this?

ajroetker commented:
@kbarber think I cleaned up for most of your comments if you wanna take another look, documentation should be much better and also added in a doc for the configuration file

kbarber commented on exe/puppet-db.cc:
```diff
@@ -40,6 +41,17 @@ main(int argc, char **argv) {
+                    "~/.puppetlabs/client-tools/puppetdb.conf"),
+                 "path to use for cli configuration")
+                ("urls", po::value<string>()->default_value(""),
+                 "urls, `http://foo:8080,http://bar:8080`, to use for puppetdb")
```
How about 'list of urls for connecting to puppetdb, urls can be comma separated. I don't know if we have a better place for examples :-).

kbarber commented on man/puppet-query.pod:
```diff
@@ -27,6 +27,29 @@ the options specific to that action.
+=item B<-c>,B<--config> <path>
+
+Overrides the path for the PuppetDB CLI config.
+Default: ~/.puppetlabs/client-tools/puppetdb.conf
```
'See `man puppetdb_conf` for more details on the contents of this file' ? Or something like that?

kbarber replied:
I think man pages also have a 'see also' section often in them. Try `man ls`.

ajroetker commented:
@kbarber you can use `nroff -u0 -Tlp -man share/man/puppetdb_conf.8 | col -x | less` to view the manpages, I adjusted everything per your suggestions

ajroetker commented:
@kbarber added a struct to hold the ssl credentials

kbarber commented:
@ajroetker hmm, do you mind if I get a tad more time to review, I'm new at reviewing C++, and I'd like to consider it again if you don't mind the wait. I want to take another look this evening, after my talkie thingy. Is that cool mate?

ajroetker commented:
@kbarber of course there's no rush here! :-)
